### PR TITLE
Knowledge Base design

### PR DESCRIPTION
**Pull request by Sizuki**
screen shot 2016-05-01 at 5 00 23 am

### Description
To implement a draft design of the Knowledge base data structure. The proposed approach is an object containing an internal `List<HornClause>`, the behaviour of the the KB allows us to `pop()` and `dequeue()` items depending on the algorithm (FC, BC).

The `HornClause` implemented contains 3 properties and a method for checking if it is `NULL` in case pop() or dequeue() returns an empty `HornClause`.

### Issues
The file parsed had to be perfectly formatted such as:
```
p2 => p3;
b => f;
a;
```
There were instances were clauses had irregular format:
```
p2=> p3;
b&e=>f;
```
Notice the lack of spaces, if in the future we are to use more propositional symbols, the question then asked would be "How do we parse this clause that is stuck together like this?".

### CODE REVIEW

File: iengine/KnowledgeBase.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
+﻿using System;
+using System.Collections.Generic;
+
+namespace iengine
+{
+  public class KnowledgeBase
+  {
+    private List<HornClause> clauseList;
+//    private HornClause[,,] clauseList2;
```
Comment (Sizuki) [inline iengine/KnowledgeBase.cs]: These are multidimensional array examples.
You can find more examples below if we are to implement them.

**Comment by Sizuki** [comment]
Oh ignore the first 3 `Console.WriteLine`'s, that's because I am outputting 3 fields but 2 of them are empty because the clause only had 1 field set.

**Comment by Sizuki** [comment]
All non-dev comments and `System.Console.WriteLines` will be removed once you give me the 👍

File: iengine/KnowledgeBase.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+//          s = clauses.TrimStart.().Split(' ');
+//          clauseList2[i][0] = s[0];
+//          clauseList2[i][1] = s[2];
+//          clauseList2[i][2] = s[1];
+//      }
+    }
+
+    private HornClause parseClause(string strClause) {
+      HornClause hc = new HornClause();
+      string[] s = strClause.Split(' ');
```
Comment (NathanDamilatis) [inline iengine/KnowledgeBase.cs]: need to confirm if irregular formatting is intended in input files

**Comment by Sizuki** [comment]
Approved by Nathan 👍

**Pull request merged**